[MRG+1] Fix KernelCenterer requires square input but doesn't error on non-square

gdex1 opened the pull request:
<!--
Thanks for contributing a pull request! Please ensure you have taken a look at
the contribution guidelines: https://github.com/scikit-learn/scikit-learn/blob/master/CONTRIBUTING.md#pull-request-checklist
-->

#### Reference Issues/PRs
Fix #14240 
<!--
Example: Fixes #1234. See also #3456.
Please use keywords (e.g., Fixes) to create link to the issues or pull requests
you resolved, so that they will automatically be closed when your pull request
is merged. See https://github.com/blog/1506-closing-issues-via-pull-requests
-->


#### What does this implement/fix? Explain your changes.
KernelCenterer.fit will now raise a ValueError when called on a non-square kernel matrix.

Added `pairwise_estimator_convert_X` to `check_fit2d_1sample` and `check_transformer_data_not_an_array` as described by https://github.com/scikit-learn/scikit-learn/issues/14240#issuecomment-511134976.

Added logical check X.T != X to assert_raises statement in `_check_transformer`.  Previously, `_check_transformer` assumed that X.T had a different number of features than X.  However, if `pairwise_estimator_convert_X` is called where transformer is pairwise, then X.T == X therefore no error ValueError will be raised so the check is invalid.  

- Adding `pairwise_estimator_convert_X` to `check_transformer_general` caused two tests to fail through  `_check_transformer` assert_raises statement that "The transformer {} does not raise an error when the number of features in transform is different from the number of features in fit."



#### Any other comments?
Fitting KernelCenterer on an nxn matrix and transforming a kxk matrix where n!=k is not currently tested by any unit test.  Doing this manually raises the error`ValueError: operands could not be broadcast together with shapes (k,k) (n,) (k,k)`.  `_check_transformer` could be modified to transpose X before `pairwise_estimator_convert_X` is called to make the test work on pairwise estimators.  KernelCenterer would need to be modified to give a more useful error message.

<!--
Please be aware that we are a loose team of volunteers so patience is
necessary; assistance handling other issues is very welcome. We value
all user contributions, no matter how minor they are. If we are slow to
review, either the pull request needs some benchmarking, tinkering,
convincing, etc. or more likely the reviewers are simply busy. In either
case, we ask for your understanding during the review process.
For more information, see our FAQ on this topic:
http://scikit-learn.org/dev/faq.html#why-is-my-pull-request-not-getting-any-attention.

Thanks for contributing!
-->

gdex1 committed:
Add raise error when KernelCenterer is used on non-square matrix

gdex1 committed:
Add pairwise_estimator_convert_X to check_fit2d_1sample and check_transformer_data_not_an_array

gdex1 committed:
Fix typo in KernelCenterer.fit

gdex1 committed:
Add test validity check to _check_transformer

gdex1 commented:
Do the proposed changes in the additional comments section seem like the appropriate ways to handle these issues?

I'm new to open source so any feedback on my solutions/code/workflow/PR is welcome. Thanks.

amueller commented on sklearn/utils/estimator_checks.py:
```diff
@@ -1059,7 +1063,8 @@ def _check_transformer(name, transformer_orig, X, y):
 
         # raises error on malformed input for transform
-        if hasattr(X, 'T') and not _safe_tags(transformer, "stateless"):
+        if hasattr(X, 'T') and not _safe_tags(transformer, "stateless") \
```
The fix I've used in other places in this file (I think) is to slice X instead of using ``X.T``. We just want something with different shape, so if you pass ``X[:, ::-1]`` instead of ``X.T`` the test works for all cases.
Otherwise you're now skipping this test for pairwise estimators.

gdex1 replied:
Thanks for the suggestion.  Would it work to use `X[:-1, :-1]` instead or does that risk messing up the test for other estimators?  I am thinking that `X[:, :-1]` will always raise a value error just because the input is not square, not because the dimensions of the input are different.

gdex1 replied:
Thanks for the suggestions.  One issue I am thinking is that for pairwise estimators, a ValueError could be triggered on `X[:, :-1]` because the input would be non-square instead of different number of dimensions.  So if X was 10x10, the estimator would throw an error on 10x9 input but possibly not 9x9 input.  Is this a problem? 

Would it make sense to do `pairwise_estimator_convert_X(X[:-1, :]).T`?  This would take a 10x10 array X to 10x9 for most estimators and 9x9 for pairwise estimators.

amueller replied:
Pairwise estimators should allow rectangular data in ``transform`` or ``predict``. I think they all do, the number of features should be the number of rows==columns in the training data.

amueller replied:
It doesn't look to me as if ``KernelCenterer`` requires square data in ``transform``.

amueller commented:
Thanks. Your proposed solution makes sense but leaves that part of the code untested as I mentioned above so I suggested an alternative.

gdex1 committed:
Improved incorrect number of features in transform raises error check

gdex1 committed:
Merge pull request #1 from scikit-learn/master

Update

gdex1 committed:
Merge branch 'master' into Fix-issue-14240

gdex1 committed:
Formatting

gdex1 committed:
Formatting 2

gdex1 committed:
Formatting 3

gdex1 committed:
Formatting 4

gdex1 committed:
Add non-square input check

gdex1 committed:
Restrict non-square input check

gdex1 committed:
Restrict non-square input check2

gdex1 commented:
I've made the change mentioned #14240 and added a test to check the change.  I am not sure if I structured the test correctly so could someone check it?

I think that possibly some other tests and errors need to be added to KernelCenterer.  Am I correct that a new issue would be a better place to discuss it?

adrinjalali commented on sklearn/utils/estimator_checks.py:
```diff
@@ -1248,6 +1259,22 @@ def check_estimators_nan_inf(name, estimator_orig):
+def check_nonsquare_error(name, estimator_orig):
+    """Test that error is thrown when non-square data provided"""
+
+    if name != "KernelCenterer":
```
Why are we skipping the test for the other ones?

gdex1 replied:
@adrinjalali. I was not sure if the test was valid for other pairwise estimators as it was causing unit tests to fail.  I was able to fix it for precomputed SVM by just changing `estimator.fit(X) -> estimator.fit(X, y)`.  

However, precomputed KNeighborRegressor failed to raise a value error on non-square data.  I tried to add a check for non-square data, but it seems other tests of KNR require non-square precomputed input matrix.  The doc says the data should be n_sample x n_sample.  What is correct? Thanks.

gdex1 committed:
Unrestrict nonsquare input check

adrinjalali commented:
It's kinda expected than when you add a test to the common tests, it detects issues with a couple of estimators that we were not detecting before. For the ones which are easy to fix, you can fix them, and for the rest, apply the changes you have done here, and we'll see together why the other tests are failing.

gdex1 committed:
Fix SVM and KNR check

gdex1 committed:
Add pairwise condition to check_regressors_no_decision_function

gdex1 committed:
Add pairwise_estimator_convert_X to check_regressors_no_decision_function

gdex1 commented:
I think everything is complete now.  In total I added:
-Value error when KernelCenterer gets non-square input
-Unit test to check pairwise estimators raise an error on non-square input
-Value error when KNeighborsRegressor(metric='precomputed') get non square input

and modified some tests that gave non-square input to pairwise estimators

adrinjalali commented on sklearn/utils/estimator_checks.py:
```diff
@@ -1248,6 +1259,19 @@ def check_estimators_nan_inf(name, estimator_orig):
+    X, y = make_blobs(n_samples=20, n_features=10)
+    estimator = clone(estimator_orig)
+
+    with assert_raises(ValueError, msg="The pairwise estimator {}"
```
please use the `pytest.raises` construct instead.

adrinjalali commented:
Otherwise LGTM!

gdex1 committed:
Change assert_raises to pytest.raises

adrinjalali commented on sklearn/utils/estimator_checks.py:
```diff
@@ -1248,6 +1260,18 @@ def check_estimators_nan_inf(name, estimator_orig):
+    X, y = make_blobs(n_samples=20, n_features=10)
+    estimator = clone(estimator_orig)
+
+    with pytest.raises(ValueError):
```
please add the `match='...'` parameter, for a more specific detection of the `ValueError`

adrinjalali commented on sklearn/utils/estimator_checks.py:
```diff
@@ -1248,6 +1260,18 @@ def check_estimators_nan_inf(name, estimator_orig):
+    estimator = clone(estimator_orig)
+
+    with pytest.raises(ValueError):
+
```
please also remove this extra line.

gdex1 committed:
Expanded pytest.raises

adrinjalali reviewed (approved):
LGTM, thanks @gdex1

jnothman commented on sklearn/utils/tests/test_estimator_checks.py:
```diff
@@ -516,6 +517,9 @@ def test_check_estimator_pairwise():
     est = SVC(kernel='precomputed')
     check_estimator(est)
 
+    est = KernelCenterer
```
This is unused

jnothman commented on sklearn/utils/tests/test_estimator_checks.py:
```diff
@@ -516,6 +517,9 @@ def test_check_estimator_pairwise():
     check_estimator(est)
 
+    est = KernelCenterer
+    check_estimator(KernelCenterer)
```
This should already be performed by test_common.py

jnothman commented on sklearn/utils/estimator_checks.py:
```diff
@@ -1059,14 +1068,17 @@ def _check_transformer(name, transformer_orig, X, y):
-        if hasattr(X, 'T') and not _safe_tags(transformer, "stateless"):
+        if hasattr(X, 'shape') and \
+           not _safe_tags(transformer, "stateless") and \
+           len(X.shape) == 2 and X.shape[1] > 1:
```
```suggestion
           X.ndim == 2 and X.shape[1] > 1:
```

jnothman reviewed (commented):
This is looking good. The nearest neighbour changes might be duplicated in #10482 but that's fine.

gdex1 committed:
Update sklearn/utils/estimator_checks.py

Co-Authored-By: Joel Nothman <joel.nothman@gmail.com>

gdex1 committed:
Remove KernelCenterer from test_check_estimator_pairwise

gdex1 committed:
Merge branch 'Fix-issue-14240' of https://github.com/gdex1/scikit-learn into Fix-issue-14240

jnothman commented:
Tests failing. I've not checked why

gdex1 committed:
Remove unused import

gdex1 commented:
@jnothman Hi. If I'm interpreting the failed tests correctly, the problem is that estimator_checks.py imports pytest for pytest.raises as described in https://github.com/scikit-learn/scikit-learn/pull/14336#discussion_r305794702.  I am not exactly sure how to fix this as other places in the repository seem to import pytest without causing an error.

jnothman commented:
Ahh right. For good or bad, we can't use pytest.raises in `estimator_checks.py` as it is considered public sklearn API and sklearn does not have a formal dependency on `pytest` (only its test suite does). So please revert to using `assert_raise*`

gdex1 committed:
Revert from pytest.raises to assert_raises

jnothman commented on sklearn/utils/estimator_checks.py:
```diff
@@ -3,6 +3,7 @@
 import sys
 import traceback
 import pickle
+import pytest
```
please remove this import

gdex1 committed:
Remove pytest import

jnothman reviewed (commented):
Please add an entry to the change log at `doc/whats_new/v0.22.rst` under the Changes to Estimator Checks section. Like the other entries there, please reference this pull request with `:pr:` and credit yourself (and other contributors if applicable) with `:user:`

gdex1 committed:
Update changelog

gdex1 commented:
@jnothman I've added document changes, but I am not completely sure if it is correct.  I also noticed that the modules are not in alphabetical order.

amueller commented:
awesome, thanks!
